### ISSUE TITLE
[Enhancement] Add better resolution for re-generated classes

### ISSUE THREAD

**Issue by sugarmanz** [intuit/hooks#16]
Currently, we have this not-so-great mechanism for handling re-generated classes.. just delete the generated files before running the compiler plugin again. This is poor, but especially for larger projects who may be spending a not insignificant portion of time generating classes. I did some investigation again and it seems as though the Arrow Proofs compiler plugin might be solving this issue and we could potentially use that as an example. Essentially, it seems as though they plug into more steps of the compiler to better inform the compiler of what the plugin is doing. 

[ProofsPlugin.kt](https://github.com/arrow-kt/arrow-meta/blob/main/plugins/proofs/proofs-plugin/src/main/kotlin/arrow/meta/plugins/proofs/ProofsPlugin.k)
[Where I think it may be handling better resolution](https://github.com/arrow-kt/arrow-meta/blob/main/plugins/proofs/proofs-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/resolve/scopes/ProofsScopeTower.kt#L40)

But, I could be off on that and this may require some additional information.

**Comment by sugarmanz** [comment]
https://kotlinlang.slack.com/archives/CJ699L62W/p1647637345993219?thread_ts=1647621597.259189&cid=CJ699L62W

Arrow Meta dropped support for generating source from typed quotes after deprecating the Proofs plugin that I linked before:
https://github.com/arrow-kt/arrow-meta/pull/969

Thus, we should look into alternatives. The simple recommendation in KSP, although that still doesn't support in-place modification of source, nor first class IDE support. Both problems with the Arrow Meta implementation, so could be sufficient. We could always build a proper compiler plugin, but that has a bit more complexity and removes the benefits of KSP.

KSP will likely be the first path of investigation.

**Comment by sugarmanz** [comment]
KSP seemed to be going well upon initial analysis. However, there seems to be an incompatibility with how we determine the type of the generated hook. We use the actual expression to determine to the types to pull out and combine with the context param:
https://github.com/intuit/hooks/blob/6e223e20b950210bdc8f16dfda17ecf876ef424d/compiler-plugin/src/main/kotlin/com/intuit/hooks/plugin/HookValidations.kt#L30

KSP doesn't seem to support this claiming it is beyond their scope:
https://github.com/google/ksp/issues/136#issuecomment-720018237

I started thinking about ways to get around this:

1. Do some reified type magic in DSL helper. I don't know if this is actually possible. Dynamically building a return type from input type parameter type parameters?
```kotlin
    protected inline fun <reified R, reified FWithoutContext : Function<R>, FWithContext : Function<R>> syncHook2(): SyncHook<FWithContext> {
        // pls ignore this 
        val fWithoutContext = FWithoutContext::class
        val fWithoutContextType = typeOf<FWithoutContext>()
        return syncHook(fWithoutContext)
    }
```

2. Instead of method helpers, try annotation helpers:
```kotlin
public abstract class Hooks : HooksDsl() {

    @Sync<(newSpeed: Int) -> Unit>()
    public abstract val accelerate: SyncHook<*>
}
```

(Which is what I ended up going with for now)
